### ISSUE TITLE
model.predict_generator() - batch_size and steps

### ISSUE THREAD

**Issue by Mahi-Mai** [keras-team/keras#11902]
I'm generating a confusion matrix to evaluate my models.  In order to do that, I need the ground truth values and predicted values for my validation dataset.

I get the true values from my ImageDataGenerator(shuffle=False).
```
validation_data_dir = '/path/val/'
nb_validation_samples = 505
img_width, img_height = 224, 224

val_datagen = ImageDataGenerator(preprocessing_function = preprocess_input)
validation_generator = val_datagen.flow_from_directory(validation_data_dir,
                                         target_size = (img_height, img_width),
                                          class_mode = "categorical",
                                          shuffle=False)

actual = validation_generator.classes
```

To get my predicted classes, I load my model and call predict_generator().
```
batch_size = 32
steps = nb_validation_samples / batch_size
predictions = model.predict_generator(val_generator, steps)

predicted_classes = convert_to_class(predictions)
```
Because predict_generator() returns a list of probabilities, convert_to_class() iterates over the list of predictions and returns a new list, where each value in the list is the index of the max number from the probability array.  Then I can use it for my confusion matrix.

So, the problem arises when I use a batch-size other than 32.  If I train my model using a different batch size, shouldn't I be running predict_generator() with the batch size I trained with?  But if I do that, I don't get the right number of samples back.  I either get too few, or too many.

How do I always ensure I'm getting predictions for all my validation samples? 

Do I not need to match  the predict_generator() batch_size with my training batch_size?

Why is 32 the magic number?

Thanks, and Happy Holidays!

**Comment by Golbstein** [comment]
you need to match the "steps" of the train\predict generator according to your batch size.

If you use different batch_size for the prediction so change the number of steps accordingly

**Comment by Dref360** [comment]
Quick tip: when using ImageDataGenerator, you should not need to specify the number of steps.

**Comment by Mahi-Mai** [comment]
@Dref360 If I don't define the steps in my predict_generator() call, I get an error requesting me to specify the steps.

**Comment by gabrieldemarmiesse** [comment]
@Mahi-Mai I think it was a keras bug but it was fixed. If you try the master branch on github this should work.

**Comment by Tan12uja** [comment]
Hello everyone, 

I am trying to get the classes for my predictions. I am using the following codes: 

actual = validation_generator.classes
predictions = model.predict_generator(val_generator, steps)
predicted_classes = convert_to_class(predictions)
The function convert_to_class does not work. What do we need to import to make it work?

Also, 
when I use 
predicted_classes = np.argmax(predictions, axis=1). It gives all the values as 0, although the command model.fit_generator gave the overall accuarcy of the model approximately 90%. My problem is a binary classification problem labelled 0 and 1 .
